crates/extensions/tedge_mqtt_ext/src/test_helpers.rs

```diff
@@ -48,14 +32,117 @@ pub async fn assert_received_includes_json<I, S>(
+    );
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum MessagePayloadMatcher {
+    StringMessage(&'static str),
+    JsonMessage(serde_json::Value),
+    Empty,
+    Skip,
+}
+
+impl From<&'static str> for MessagePayloadMatcher {
+    fn from(value: &'static str) -> Self {
+        MessagePayloadMatcher::StringMessage(value)
+    }
+}
+
+impl From<serde_json::Value> for MessagePayloadMatcher {
+    fn from(value: serde_json::Value) -> Self {
+        MessagePayloadMatcher::JsonMessage(value)
+    }
+}
+
+pub fn assert_messages_matching<'a, M, I>(messages: M, expected: I)
+where
+    M: IntoIterator<Item = &'a Message>,
+    I: IntoIterator<Item = (&'static str, MessagePayloadMatcher)>,
+{
+    let mut messages_iter = messages.into_iter();
+    let mut expected_iter = expected.into_iter();
+    loop {
+        match (messages_iter.next(), expected_iter.next()) {
+            (Some(message), Some(expected_msg)) => {
+                let message_topic = &message.topic.name;
+                let expected_topic = expected_msg.0;
+                match expected_msg.1 {
+                    MessagePayloadMatcher::StringMessage(str_payload) => {
+                        assert_message_contains_str(message, (expected_topic, str_payload))
+                    }
+                    MessagePayloadMatcher::JsonMessage(json_payload) => {
+                        assert_message_includes_json(message, (expected_topic, json_payload))
+                    }
+                    MessagePayloadMatcher::Empty => {
+                        assert_eq!(
+                            message_topic, expected_topic,
+                            "Received message on topic: {} instead of {}",
+                            message_topic, expected_topic
+                        );
+                        assert!(
+                            message.payload_bytes().is_empty(),
+                            "Received non-empty payload while expecting empty payload on {}",
+                            message_topic
+                        )
+                    }
+                    MessagePayloadMatcher::Skip => {
+                        assert_eq!(
+                            message_topic, expected_topic,
+                            "Received message on topic: {} instead of {}",
+                            message_topic, expected_topic
+                        );
+                        // Skipping payload validation
+                    }
+                }
+            }
+            (None, Some(expected_msg)) => {
+                panic!(
+                    "Input messages exhausted while expecting message on topic: {:?}",
+                    expected_msg.0
+                )
+            }
+            (Some(message), None) => {
+                panic!(
+                    "Additional message received than expected on topic: {:?}",
+                    message.topic.name
+                )
+            }
+            _ => return,
+        }
     }
```

Comment:
This combination of `loop` and `match` feels clunky. I think that a lot of the nesting and complexity can go away with something like:

```suggestion
    let mut messages = messages.into_iter();
    for expected_msg in expected {
        let Some(message) = messages.next() else {
            panic!(
                "Input messages exhausted while expecting message on topic: {:?}",
                expected_msg.0
            )
        };
        let message_topic = &message.topic.name;
        let expected_topic = expected_msg.0;
        match expected_msg.1 {
            MessagePayloadMatcher::StringMessage(str_payload) => {
                assert_message_contains_str(message, (expected_topic, str_payload))
            }
            MessagePayloadMatcher::JsonMessage(json_payload) => {
                assert_message_includes_json(message, (expected_topic, json_payload))
            }
            MessagePayloadMatcher::Empty => {
                assert_eq!(
                    message_topic, expected_topic,
                    "Received message on topic: {} instead of {}",
                    message_topic, expected_topic
                );
                assert!(
                    message.payload_bytes().is_empty(),
                    "Received non-empty payload while expecting empty payload on {}",
                    message_topic
                )
            }
            MessagePayloadMatcher::Skip => {
                assert_eq!(
                    message_topic, expected_topic,
                    "Received message on topic: {} instead of {}",
                    message_topic, expected_topic
                );
                // Skipping payload validation
            }
        }
    }
    
    if let Some(message) = messages.next() {
        panic!(
            "Additional message received without expectation on topic: {:?}",
            message.topic.name
        );
    }
```

Reply:
Just a matter of style, I guess. For some reason, I still have a little preference for the existing style where all the match combinations are laid out directly and even the `panic!` cases are part of those cases. But, if other also prefer this style with less nesting, I'm happy to change it.

Reply:
Personally, I find it really quite difficult to read (and modify) "diagonal" code (e.g. where the 8 lines starting from `loop {` cover 6 different indentation levels). In this case, what's going on is very simple, so it doesn't a huge difference. If you want the `loop` because it's explicit, it would make sense to replace the catch-all (i.e. `_ => ...` branch) with `(None, None) => ...` as that's what we care about here, and it reduces the mental overhead of having to remember exactly what we have and haven't handled (and catches potential future errors if someone decides to delete them). Also, the distinction here is pretty obvious, but in a long function, it can be quite difficult to distuingish between the various cases that arise in nested matches.